extension/src/testReportProvider.ts

```diff
@@ -1,51 +1,180 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license.
 
-import { TestLevel, TestSuite } from "./protocols";
+import { TestLevel, TestStatus, TestSuite } from "./protocols";
 import { TestResourceManager } from "./testResourceManager";
 
 import * as os from 'os';
-import { TextDocumentContentProvider, Uri } from "vscode";
+import * as path from 'path';
+import { window, ExtensionContext, TextDocumentContentProvider, Uri } from "vscode";
 
 export class TestReportProvider implements TextDocumentContentProvider {
 
     public static scheme = 'test-report';
 
-    constructor(private _testResourceProvider: TestResourceManager) {
+    constructor(private _context: ExtensionContext, private _testResourceProvider: TestResourceManager) {
     }
 
     public provideTextDocumentContent(uri: Uri): string {
         const [target, test] = decodeTestSuite(uri);
         const testsContainedInFile = this._testResourceProvider.getTests(target);
         if (!testsContainedInFile) {
-            return `No tests in the uri: ${uri}. Shouldn\'t happen, please report us a bug.`;
+            return this.errorSnippet(`No tests in the uri: ${uri}. Shouldn\'t happen, please report us a bug.`);
         }
         if (testsContainedInFile.dirty) {
-            return 'Test file has been changed, try open the report again...';
+            return this.errorSnippet('Test file has been changed, try open the report again...');
         }
         const testMap: Map<string, TestSuite> = new Map(testsContainedInFile.tests.map((t): [string, TestSuite] => [t.test, t]));
         const matchedTest = testMap.get(test);
         if (!matchedTest) {
-            return 'No matched test found in the test storage. Shouldn\'t happen, please report us a bug.';
+            return this.errorSnippet('No matched test found in the test storage. Shouldn\'t happen, please report us a bug.');
         }
-        return this.getTestReport(matchedTest);
+        return this.reportSnippet(matchedTest);
     }
 
-    // to-do: better formatting
-    private getTestReport(test: TestSuite): string {
-        let report = test.test + ':' + os.EOL;
-        if (!test.result) {
-            return report + "Not run...";
-        }
-        report += JSON.stringify(test.result, null, 4);
-        if (test.level === TestLevel.Method) {
-            return report;
-        }
-        report += os.EOL;
-        for (const child of test.children) {
-            report += this.getTestReport(child) + os.EOL;
+    private errorSnippet(error: string): string {
+        return `
+            <body>

```

Comment:
Hard-coding HTML is never good. Please use a template engine. In this way, the HTML can be extracted as a separate template file and be maintained separately.